pkg/shared/utils/containerRootfs/fs.go

```diff
@@ -0,0 +1,166 @@
+	}
+	defer image.Cleanup()
+
+	err = image.SquashedTree().Walk(func(path file.Path, f filenode.FileNode) error {
+		target := filepath.Join(dest, string(path))
+
+		switch f.FileType {
+		case file.TypeDirectory:
+			if err := os.MkdirAll(target, 0755); err != nil {
+				return err
+			}
+		case file.TypeRegular:
+			indexEntry, err := image.FileCatalog.Get(*f.Reference)
+			if err != nil {
+				return fmt.Errorf("unable to get index entry for file: %w", err)
+			}
+
+			output, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, indexEntry.Mode())
+			if err != nil {
+				return fmt.Errorf("unable to open file %s: %w", target, err)
+			}
+
+			content, err := image.OpenReference(*f.Reference)
+			if err != nil {
+				return fmt.Errorf("unable to open path from squash: %w", err)
+			}
+
+			numBytes, err := io.Copy(output, io.LimitReader(content, perFileReadLimit))
+			if numBytes >= perFileReadLimit || errors.Is(err, io.EOF) {
+				return fmt.Errorf("zip read limit hit (potential decompression bomb attack)")
+			}
+			if err != nil {
+				return fmt.Errorf("unable to copy file: %w", err)
+			}
+
+			err = output.Close()
+			if err != nil {
+				// TODO log the error
+			}
+		}
+
+		return nil
+	}, nil)
+	if err != nil {
+		return fmt.Errorf("unable to walk squashed tree from container: %w", err)
+	}
+
+	return nil
+}
+
+type Rootfs interface {
+	Dir() string
+	Cleanup() error
+}
+
+func ToTempDirectory(ctx context.Context, src string) (Rootfs, error) {
+	tdrf := &TempDirRootfs{}
+
+	tmpdir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return tdrf, fmt.Errorf("unable to create temp directory: %w", err)
+	}
+	tdrf.dir = tmpdir
+
+	err = ToDirectory(ctx, src, tmpdir)
+	if err != nil {
+		return tdrf, fmt.Errorf("unable output squashed image to directory: %w", err)
+	}
+
+	return tdrf, nil
+}
+
+type TempDirRootfs struct {
+	dir string
+}
+
+func (tdrf *TempDirRootfs) Dir() string {
+	return tdrf.dir
+}
+
+func (tdrf *TempDirRootfs) Cleanup() error {
+	var errs []error
+
+	if tdrf.dir != "" {
+		err := os.RemoveAll(tdrf.dir)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to remove temp directory: %w", err))
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+type cachedRootfs struct {
+	complete chan struct{}
+	err      error
+	rootfs   Rootfs
+}
+
+var cacheMutex sync.RWMutex
+var cache map[string]*cachedRootfs = map[string]*cachedRootfs{}
+
+func (crfs *cachedRootfs) Dir() string {
+	return crfs.rootfs.Dir()
+}
+
+func (crfs *cachedRootfs) Cleanup() error {
+	return nil
+}
+
+func ToCachedTempDirectory(ctx context.Context, src string) (Rootfs, error) {
+	cacheMutex.Lock()
+	entry, ok := cache[src]
+	if ok {
+		cacheMutex.Unlock()
+		<-entry.complete
+		return entry, entry.err
+	}
+	cachedEntry := &cachedRootfs{
+		complete: make(chan struct{}),
+		err:      nil,
+	}
+	cache[src] = cachedEntry
+	cacheMutex.Unlock()
+
+	rootfs, err := ToTempDirectory(ctx, src)
+	cachedEntry.err = err
```

Comment:
If one call to `ToCachedTempDirectory` failed for this `src` there is no way to try it again?

Reply:
not right now as it keeps the in progress wait case and the complete wait case the same so this code is simpler. I can certainly update it so that if we receive a new call to ToCachedTempDirectory after it has completed with error that it tries again. WDYT?